### PR TITLE
Async Processors and Final Pipelines are Broken

### PR DESCRIPTION
**Pull request by original-brownbear**
There was an obvious race here where async processor and final pipeline will
run concurrently (or the final pipeline runs multiple times in from the while loop).

relates #52339 (fixes one failure scenario here but since the failure also occurred in 7.10.x not all of them)

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/index/FinalPipelineIT.java [hunk 1]
```diff
@@ -357,9 +358,16 @@ public void testHighOrderFinalPipelinePreferred() throws IOException {
...
+                        public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
+                            randomFrom(parameters.genericExecutor, Runnable::run).accept(() -> {
+                                ingestDocument.setFieldValue("default", true);
+                                handler.accept(ingestDocument, null);
+                            });
+                        }
+
+                        @Override
+                        public IngestDocument execute(IngestDocument ingestDocument) {
+                            throw new UnsupportedOperationException();
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/index/FinalPipelineIT.java:370]: Maybe this?

```suggestion
                            throw new AssertionError("should not be called");
```

**Review by probakowski — APPROVED** [review]
LGTM, thanks for fixing this @original-brownbear!
I wondered if we can hit stack overflow in case of sync processors (we do recurse in that case) but I think it's not a problem here as `it` should have at most 2 items in it

File: server/src/internalClusterTest/java/org/elasticsearch/index/FinalPipelineIT.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/index/FinalPipelineIT.java]: ++

**Comment by original-brownbear** [comment]
Thanks @probakowski !

**Pull request merged**